Add support for Rook backend

BlaineEXE opened the pull request:
This  makes the 'ses' plugin a small file that
calls out to larger plugin 'resource' files. The new idea is to make
more use of the /usr/lib/supportconfig/resources directory for
containing resources, especially since there is a lot of code for
collecting Ceph logs. Most of the code will live in resources and be
called/loaded as-needed.

This will mean that the build service package needs to be built slightly
differently. Instead of placing the 'ses-rpm-list' file into the plugin
resources directory, the whole new 'ses-resources' directory here in
source will need to be installed to /usr/lib/supportconfig/resources/ses

I teased out the preexisting validate_rpm_if_installed() function into a
helpers.sh in the ses-resources/ dir to prove out the above change.
Additionally, most of the current validation code (almost all) has been
moved to the ses-resources/cephadm script.

Between cephadm and rook, collect information about Ceph daemons using
'ceph daemon <daemon>' commands (and other related code) in a shared
library.

Add a helper function to collect info from `ceph` CLI. In Rook, use the
collector helper pod, and in cephadm, just call the function directly.
The helper function takes as an argument the shell command where it can
reach the `ceph` CLI.

I have made some changes to Tim's original design:
- added more plugin_message's in the section
- added a few additional outputs to logs
- added a 'timeout $((CT * 2))' to all commands
- use json output and `jq` instead of sed for parsing CLI outputs
  - except for replacing values with $CENSORED output

Try to connect via cephadm and via kubectl to determine if Ceph is
running via cephadm or if Rook is configured to find the Ceph/SES
cluster to log info from.

Also fix shellcheck warnings.

Remove an unused 'container_id' in the loop that collects daemon
information in the cephadm script.

Collect resource details from the kube-system namespace to help verify
whether issues are a result of an underlying Kubernetes issue (like with
CNI failures). This shouldn't be necessary for CaaSP but couled be
useful for non-CaaSP clusters.

Symlink from the Ceph daemon log dir to target appropriate pod logs so
users will have a similar layout of Ceph daemon logs for both Rook and
cephadm backends.

Signed-off-by: Blaine Gardner <blaine.gardner@suse.com>

BlaineEXE committed:
Add support for rook backend

This  makes the 'ses' plugin a small file that
calls out to larger plugin 'resource' files. The new idea is to make
more use of the /usr/lib/supportconfig/resources directory for
containing resources, especially since there is a lot of code for
collecting Ceph logs. Most of the code will live in resources and be
called/loaded as-needed.

This will mean that the build service package needs to be built slightly
differently. Instead of placing the 'ses-rpm-list' file into the plugin
resources directory, the whole new 'ses-resources' directory here in
source will need to be installed to /usr/lib/supportconfig/resources/ses

I teased out the preexisting validate_rpm_if_installed() function into a
helpers.sh in the ses-resources/ dir to prove out the above change.
Additionally, most of the current validation code (almost all) has been
moved to the ses-resources/cephadm script.

Between cephadm and rook, collect information about Ceph daemons using
'ceph daemon <daemon>' commands (and other related code) in a shared
library.

Add a helper function to collect info from `ceph` CLI. In Rook, use the
collector helper pod, and in cephadm, just call the function directly.
The helper function takes as an argument the shell command where it can
reach the `ceph` CLI.

I have made some changes to Tim's original design:
- added more plugin_message's in the section
- added a few additional outputs to logs
- added a 'timeout $((CT * 2))' to all commands
- use json output and `jq` instead of sed for parsing CLI outputs
  - except for replacing values with $CENSORED output

Try to connect via cephadm and via kubectl to determine if Ceph is
running via cephadm or if Rook is configured to find the Ceph/SES
cluster to log info from.

Also fix shellcheck warnings.

Remove an unused 'container_id' in the loop that collects daemon
information in the cephadm script.

Collect resource details from the kube-system namespace to help verify
whether issues are a result of an underlying Kubernetes issue (like with
CNI failures). This shouldn't be necessary for CaaSP but couled be
useful for non-CaaSP clusters.

Symlink from the Ceph daemon log dir to target appropriate pod logs so
users will have a similar layout of Ceph daemon logs for both Rook and
cephadm backends.

Update the man page.

Signed-off-by: Blaine Gardner <blaine.gardner@suse.com>

tserong commented on ses-resources/cephadm:
```diff
@@ -0,0 +1,200 @@
+
+
+#############################################################
+rpm_list="$SES_SES_RESOURCES_DIR"/rpm-list
```
Should be `rpm_list="$SES_RESOURCES_DIR"/rpm-list` (i.e. remove the extra "SES")

tserong commented on ses-resources/cephadm:
```diff
@@ -0,0 +1,200 @@
+for thisrpm in $(cat "$rpm_list"); do
+    validate_rpm_if_installed "$thisrpm"
+done
+
```
The entire block above here belongs in the main `ses` script, as rpm validation should happen for both the cephadm and rook cases.

BlaineEXE replied:
Are we sure that's the case? Kubernetes nodes won't need the same RPMs installed. 

I will change this, because I do see that it could be useful to validate RPMs on k8s nodes sometimes, but I think most of the time, the RPM validation won't tell very much about Rook clusters.

tserong replied:
You're right, the RPM validation won't tell us much about the rook side of thinks, but, the RPMs in this list are all the RPMs we ship on the SES media, so regardless of whether we're using k8s or cephadm, if any of them are installed on the host, we want to know about them.

tserong commented on ses-resources/cephadm:
```diff
@@ -0,0 +1,200 @@
+#############################################################
+section_header "Ceph cluster status"
+
+mkdir $LOG/ceph
```
This line should be removed (it's already handled on line 26 of the main ses script, and including it here results in an error message, as the directory already exists)

tserong commented on ses-resources/cephadm:
```diff
@@ -0,0 +1,200 @@
+    # timeout status commands after 5 seconds
+    CT=5
+
+    collect_info_from_cli "$CEPH_SHELL"
```
Should be `collect_info_from_ceph_cli "$CEPH_SHELL"`

tserong reviewed (commented):
I've done some testing of the cephadm case, with changes applied as suggested below, and all seems to work well (see https://build.suse.de/package/show/home:tserong:branches:SUSE:SLE-15-SP2:Update:Products:SES7/supportutils-plugin-ses for the packaged version with spec file update to handle the split out resources).

I'm wondering if we should stick a ".rc" suffix on each of the scripts in the resources directory, just for consistency with the naming of such files in the main supportutils package (after all, these are all `source`d, not meant for direct execution by humans). That's really just cosmetic though.

Code-wise, I like that things are split out into into a few separate files, but I've just now realised that there's one thing we *lose* by doing this: supportconfig runs `cat -n $PLUGIN_BIN` for each plugin, so ordinarily the source of each suportconfig plugin is appended to the plugin log file in the supportconfig tarball, so you can see exactly what was executed. Now, with most of the functionality split out into separate files, all we see in the plugin log is effectively a case statement, and none of the actual guts of the code. I'm not sure if this is actually a problem, as we ship supportutils-plugin-ses with SES, and we know what's in there anyway. I suspect `cat`ing the plugin source is more interesting if users have written their own plugins. Still, I'd like to get another opinion on this. @g23guy, are you able to comment? Please also see the spec file at https://build.suse.de/package/view_file/home:tserong:branches:SUSE:SLE-15-SP2:Update:Products:SES7/supportutils-plugin-ses/supportutils-plugin-ses.spec?expand=1 for an idea of how we're looking at packaging this thing now.

BlaineEXE commented:
@tserong : I changed them to .rc files in my cleanup PR: https://github.com/SUSE/supportutils-plugin-ses/pull/46

BlaineEXE commented:
@tserong and @g23guy
I didn't realize the plugin script would be `cat`ed. Most of the other plugins I see are pretty short. Ceph's is a long plugin, even without Rook stuff ongoing. Maybe it could be a good idea to have the `ses` script manually `cat` the source files out itself?

g23guy commented:
> @tserong and @g23guy
> I didn't realize the plugin script would be `cat`ed. Most of the other plugins I see are pretty short. Ceph's is a long plugin, even without Rook stuff ongoing. Maybe it could be a good idea to have the `ses` script manually `cat` the source files out itself?

My intent was to capture the source code to help debug.

BlaineEXE commented:
@tserong and @g23guy I added something in https://github.com/SUSE/supportutils-plugin-ses/pull/46/commits/7a9f6f427e1ae90fdfb0db7d21c49ec62bec294c part of #46

tserong commented:
I'm finding I'm having difficulty reviewing this, with subsequent changes being added to a separate PR (#46). Can you please merge the changes from #46 back into this branch, so we can just continue discussion here in #45?

Also, it looks like #46 is missing at least one fix from this branch (`collect_info_from_ceph_cli` vs `collect_info_from_cli` in cephadm.rc).

Re: catting the source, 7a9f6f4 doesn't seem to work because nothing actually ever creates the LOGPLUGIN directory, so when I run it, I get "/usr/lib/supportconfig/plugins/ses: line 162: /var/log/scc_node1_200827_1315/ceph/supportutils-plugin/index-of-files: No such file or directory". I'd suggest to not bother putting all this stuff in a separate directory, just have it output each file directly to the plugin-ses.txt file. Invoking `pconf_files $SES_RESOURCESD_DIR/*` is probably enough to do this (but this needs to be tested). There should be no need for the extra block with an new section header and plugin_message calls. Sure, `pconf_files` will list each file under headings that say "Configuration File", but that's fine, it'll do the job, and if this is the *last* thing executed by the plugin, it'll be directly before the automatic `cat` of the main plugin file, so it'll all make some sort of sense to read.

BlaineEXE commented:
I don't know what the `plugin-ses.txt` file is... I'm guessing that's just stdout from this script? I added the changes here and removed them from the cleanup pr... eventually squashed everything to make it through rebase issues.

tserong commented:
> I don't know what the `plugin-ses.txt` file is... I'm guessing that's just stdout from this script? I

Correct, sorry for not being clearer about that. This is looking good, and the test package I built works fine with just two small remaining kinks:

1) A backtick has somehow landed in a `jq` invocation in cephadm.rc, which is fixable with this:

```diff --git a/ses-resources/cephadm.rc b/ses-resources/cephadm.rc
index 72dcfee..f223fa5 100755
--- a/ses-resources/cephadm.rc
+++ b/ses-resources/cephadm.rc
@@ -74,7 +74,7 @@ if [ -d /var/log/ceph ]; then
     plugin_message "Ceph logs copied to ceph/log subdirectory"
 fi
 
-cephadm ls | jq -r '.[] | "\(.fsid) \(.n`ame)"' |
+cephadm ls | jq -r '.[] | "\(.fsid) \(.name)"' |
     while read fsid name ; do
         mkdir -p "$LOGCEPHLOG"
         # Individual daemon logs from journalctl named without the leading "ceph-"
```

2) ses-resources/cephadm.sh needs deleting (I assume that's a rebase artifact, as it's now cephadm.rc)

Thanks!
